xwiki-platform-core/xwiki-platform-repository/xwiki-platform-repository-server-api/src/main/java/org/xwiki/repository/internal/RepositoryManager.java

```diff
@@ -1102,4 +1198,38 @@ private boolean updateCollection(BaseObject extensionObject, String fieldName, C
 
         return needSave;
     }
+
+    public BaseObject getExtensionVersionObject(XWikiDocument extensionDocument, String version) {
+        return getExtensionVersionObject(extensionDocument, version, true);
+    }
+
+    public BaseObject getExtensionVersionObject(XWikiDocument extensionDocument, String version, boolean allowProxying)
+    {
+        if (version == null) {
+            List<BaseObject> objects =
+                    extensionDocument.getXObjects(XWikiRepositoryModel.EXTENSIONVERSION_CLASSREFERENCE);
+
+            if (objects == null || objects.isEmpty()) {
+                return null;
+            } else {
+                return objects.get(objects.size() - 1);
+            }
+        }
+
+        BaseObject extensionVersionObject =
+                extensionDocument.getObject(XWikiRepositoryModel.EXTENSIONVERSION_CLASSNAME, "version", version, false);
+
+        if( extensionVersionObject == null && allowProxying && isVersionProxyingEnabled(extensionDocument) ){
+             //no ExtensionVersionClass object so we need to create such object temporarily
```

Comment:
> Maybe it's not such a big drawback?

Having a document which is different from the cache and the database can cause all sorts of issues. Here is an example: you do a REST request which add an object in the XWikiDocument, if you then call import it will try to delete that object (since it's supposed to proxy this version) that does and will fail to save. Another issue if that you if someone modify the document for any reason (add some documentation) it will actually save the object in the database and the history won't make any sense ("hey I did not added this object !"), etc.

At the very least you should clone the document before adding an object it and put a FIXME comment to improve this hack with something cleaner in the future.